File: 1036-rotting-oranges/1036-rotting-oranges.cpp
class Solution {
public:
int orangesRotting(vector<vector<int>>& grid) {
    int m = grid.size();
    int n = grid[0].size();
    queue<pair<pair<int,int>,int>> q;
    vector<vector<int>> visited(m,vector<int>(n,0));
    int cntFresh = 0;
    int newRotten = 0;
    for(int i = 0; i < m; i++){
        for(int j = 0; j < n; j++){
            if(grid[i][j] == 2){
                q.push({{i,j},0});
                visited[i][j] = 2;
            }
            if(grid[i][j] == 1)cntFresh++;
        }
    }

    int time = 0;

    while(!q.empty()){
        pair<pair<int,int>,int> front = q.front();
        int r = front.first.first;
        int c = front.first.second;
        int t = front.second;
        q.pop();

        time = t;

        int drow[] = {1,0,-1,0};
        int dcol[] = {0,1,0,-1};

        for(int i = 0; i < 4; i++){
            int nrow = r + drow[i];
            int ncol = c + dcol[i];

            if(nrow >= 0 && nrow < m && ncol >= 0 && ncol < n && visited[nrow][ncol] != 2 && grid[nrow][ncol] == 1){
                q.push({{nrow,ncol},1+t});
                visited[nrow][ncol] = 2;
                newRotten++;
            }
        }
    }
    if(cntFresh != newRotten)return -1;
    return time;
    }
};

File: 1036-rotting-oranges/1036-rotting-oranges.java
import java.util.AbstractMap.SimpleEntry;
class Solution {
    public int orangesRotting(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;

        Queue<SimpleEntry<SimpleEntry<Integer,Integer>, Integer>> q = new LinkedList<>();
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == 2){
                    SimpleEntry<Integer,Integer> coord = new SimpleEntry<>(i,j);
                    q.add(new SimpleEntry<>(coord, 0)); // add (i,j) with time 0
                }
            }
        }

        int minTime = 0;
        int[][] del = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        
        while(!q.isEmpty()){
            SimpleEntry<SimpleEntry<Integer,Integer>, Integer> front = q.remove();
            int i = front.getKey().getKey();
            int j = front.getKey().getValue();
            int time = front.getValue();
            
            minTime = Math.max(minTime, time);

            for(int k = 0; k < 4; k++){
                int nrow = i + del[k][0];
                int ncol = j + del[k][1];

                if(nrow >= m || ncol >= n || nrow < 0 || ncol < 0 || grid[nrow][ncol] != 1) continue;

                grid[nrow][ncol] = 2;
                SimpleEntry<Integer,Integer> coord = new SimpleEntry<>(nrow,ncol);
                q.add(new SimpleEntry<>(coord, time + 1));
            }
        }

        // Check for any remaining fresh oranges
        for(int[] row : grid){
            for(int cell : row){
                if(cell == 1) return -1;
            }
        }

        return minTime;
    }
}





// class Solution {
//     public int orangesRotting(int[][] grid) {
//         int m = grid.length;
//         int n = grid[0].length;

//         Queue<SimpleEntry<SimpleEntry<Integer,Integer>,Integer>>q = new LinkedList<>();
//         for(int i = 0; i < m; i++){
//             for(int j = 0; j < n; j++){
//                 if(grid[i][j] == 2){
//                     SimpleEntry<Integer,Integer>s = new SimpleEntry<>(i,j);
//                     SimpleEntry<Integer,Integer>p = new SimpleEntry<>(s,0);
//                     q.add(p);
//                 }
//             }
//         }

//         int minTime = 0;

//         while(!q.isEmpty()){
//             SimpleEntry<SimpleEntry<Integer,Integer>,Integer> front = q.remove();
//             int i = front.getKey().getKey();
//             int j = front.getKey().getValue();
//             int time = front.getValue(); // time at which orange becomes rotten
            
//             minTime = Math.max(minTime,time); // maxtime taken by a orange to rotten -> final time

//             int[][] del = {{0, 1}, {0, -1},{1, 0},{-1, 0}};
//             for(int k = 0; k < 4; k++){
//                 int nrow = i + del[k][0];
//                 int ncol = j + del[k][1];

//                 if(nrow == m || ncol == n || nrow < 0 || ncol < 0 || grid[nrow][ncol] != 1) continue;

//                 grid[nrow][ncol] = 2;
//                 SimpleEntry<Integer,Integer>s = new SimpleEntry<>(nrow,ncol);
//                 SimpleEntry<Integer,Integer>p = new SimpleEntry<>(s,time+1);

//                 q.add(p);
//             }
//         }

//         for(int[] row : grid){
//             for(int cell : row){
//                 if(cell == 1)return -1;
//             }
//         }

//         return minTime;
//     }
// }